rclcpp/test/rclcpp/test_qos_event.cpp

```diff
@@ -222,3 +229,97 @@ TEST_F(TestQosEvent, test_default_incompatible_qos_callbacks)
 
   rcutils_logging_set_output_handler(original_output_handler);
 }
+
+TEST_F(TestQosEvent, construct_destruct_rcl_error) {
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>(topic_name, 10);
+  auto rcl_handle = publisher->get_publisher_handle();
+
+  bool handler_callback_executed = false;
+  // This callback requires some type of parameter, but it could be anything
+  auto callback = [&handler_callback_executed](int) {handler_callback_executed = true;};
+  rcl_publisher_event_type_t event_type = RCL_PUBLISHER_OFFERED_DEADLINE_MISSED;
+
+  {
+    // Logs error and returns
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rclcpp", rcl_publisher_event_init, RCL_RET_ERROR);
+
+    auto throwing_statement = [&callback, &rcl_handle, &event_type]() {
+        // reset() is not needed for the exception, but it handles unused return value warning
+        std::make_shared<
+          rclcpp::QOSEventHandler<decltype(callback), decltype(rcl_handle)>>(
+          callback, rcl_publisher_event_init, rcl_handle, event_type).reset();
+      };
+    // This is done through a lambda because the compiler is having trouble parsing the templated
+    // function inside a macro.
+    EXPECT_THROW(throwing_statement(), rclcpp::exceptions::RCLError);
+  }
+
+  {
+    // Logs error and returns
+    auto mock = mocking_utils::patch_and_return(
```

Comment:
Why not `inject_on_return` as it is a `fini` method?

Reply:
Done